Add Swin Transformer Backbone

AdonaiVera opened the pull request:
## Add Swin Transformer Backbone
This PR adds `SwinTransformerBackbone`, based on the paper [Swin Transformer: Hierarchical Vision Transformer using Shifted Windows](https://arxiv.org/abs/2103.14030). It is the first step to support **Grounding DINO** in `keras-hub`, where Swin is the main image encoder.

### Related Work

This idea has been discussed in previous threads:
* [`keras-cv` issue #2114](https://github.com/keras-team/keras-cv/issues/2114) — community request to support Grounding DINO, which requires Swin.
* [`keras-hub` issue #2117](https://github.com/keras-team/keras-hub/issues/2117) — A request to support Swin-UNETR, a model originally designed for 3D medical image segmentation using Swin Transformers for effective feature extraction. 

###  Included in this PR
* `SwinTransformerBackbone` model
* Core layers (`PatchEmbedding`, `WindowAttention`, etc.)
* Unit tests
* (Presets will be added later)

### 🧪 Current Status

This PR is still a **draft**. I’m finishing:

* Fixing some issues with `float16` and `float32`
* Preparing a **Colab notebook** to:
  * Convert pretrained weights from the original repo
  * Compare the outputs with the original model

Let me know if this direction makes sense, or if you have any comments or suggestions. Thanks!

AdonaiVera committed:
Add backbone, layer and test of swin transform - image encoder

AdonaiVera committed:
First iteration of swin transformer layers in keras hub models

AdonaiVera committed:
Deleted Swin Transformer files, remove typo

AdonaiVera committed:
Fixed dtype issue

AdonaiVera committed:
Fix formatting and undefined variable errors

abheesht17 commented:
@AdonaiVera - are you still working on this?

AdonaiVera committed:
Added missing dropout, and initialize it directly as a Keras tensor

AdonaiVera committed:
Added normalization init and name

innat commented:
@AdonaiVera 
You can take a look into this [implementation ](https://github.com/sayakpaul/swin-transformers-tf)(2d-swin), it reproduced original implementaiton (sort of - though there are some [limitaiton](https://github.com/sayakpaul/swin-transformers-tf/issues/9), which you might care to fix).

AdonaiVera committed:
keras_hub/src/models/swin_transformer/swin_transformer_layers.py

AdonaiVera committed:
Added explicit padding mode for cross-backend compatibility.

AdonaiVera commented:
Hi Keras Team! 

Sorry for the late reply. It’s been almost 6 months without activity. I joined a new job and was very, very busy. I only had some time during the weekend to work on this again. I hope this model is still relevant, I think it is, since it enables other models that use this backbone.

## Update:
1.  I finished the final adjustments, and now the code replicates the Hugging Face structure. You can find the comparison in this [Colab](https://colab.research.google.com/drive/1hw-HKrEZT4nqoN-poY6G2NxE1xrw6t04?usp=sharing).

2. @innat I validated it against the 2D Swin implementation and we are aligned. I also checked the constraints. This implementation removes the hard-coded resolution limitation present in previous TensorFlow versions. The model now supports arbitrary square input sizes (not limited to 224×224 or 384×384). Attention masks are computed dynamically at runtime, keeping the same parameter count across different input sizes. I think we are good on this side.

3. @divyashreepathihalli  I validated the Gemini feedback. Out of the five Gemini CLI suggestions, three were already correctly implemented (using ops.zeros, ops.cast for division, and no reshape_and_norm loop issue). We implemented the one valuable suggestion: adding an explicit mode="constant" to the padding operation for better cross-backend compatibility. We skipped one low-priority micro-optimization related to attention mask creation.

## Next steps:
If everything looks good, I’ll:
1. create the checkpoint conversion script
2. add the preset
3. upload the weights to Kaggle

For now,  I’ll wait for feedback before moving forward.

Thanks in advance, and sorry again for the delay 🙏

AdonaiVera committed:
Fix Swin Transformer: JAX compatibility and code quality

AdonaiVera committed:
Add Swin Transformer presets and conversion script

AdonaiVera committed:
Merge branch 'keras-team:master' into master

AdonaiVera committed:
Add SwinTransformerBackbone to public API exports

innat commented:
@divyashreepathihalli Hi, could you share some insight into why vision models seem to be receiving less focus compared to LLM-based contributions? I’m asking because Keras Hub currently has many LLM models, while vision models appear to be lower priority.

I’m also curious because some of the existing vision models in Keras Hub (for example, YOLOv8) still have several unresolved bugs. I may be overlooking something, and I understand there may be an internal prioritization or roadmap. If such a priority list exists, is it publicly available? That could help contributors align their efforts more effectively and reduce uncertainty when contributing.

Thanks for the clarification.

divyashreepathihalli commented:
@sachinprasadhs can you please take a look? 

@innat - we have the same priorities for vision and language models. I am sorry this PR has slipped through the cracks, feel free to tag us if we are unresponsive.

sachinprasadhs reviewed (commented):
Thanks for the PR. 
I have done one pass to the implementation and made some comments. please address it and mark it as resolved once you have addressed it.
below are my general comments.

1. The validation error tolerance is big , decrease the tolerance to match at least 1e-4, since this model will be considered as encoder for other models, we need to make sure the model accuracy is upto the mark.
2. Does this HF model comes with the classification head weights?
3. Can we include the classifier, prepropcessor for ayone who wants to finetune the model.

sachinprasadhs commented:
@gemini-code-assist review

AdonaiVera commented on keras_hub/src/models/swin_transformer/swin_transformer_backbone_test.py:
```diff
@@ -0,0 +1,58 @@
+            "window_size": 7,
+            "mlp_ratio": 4.0,
+        }
+        self.input_data = ops.ones((1, 224, 224, 3))
```
Yes, I changed to a smaller config `(32x32 input, 2 stages, embed_dim=32)` and its much faster, 2.75s vs the previous 9.31s.

AdonaiVera commented on keras_hub/src/models/swin_transformer/swin_transformer_backbone.py:
```diff
@@ -0,0 +1,194 @@
+        drop_path=0.1,
+        patch_norm=True,
+        data_format="channels_last",
+        dtype=None,
```
Updated. Made the core architecture parameters `(embed_dim, depths, num_heads, window_size)` required (no defaults), following the same convention as ViT and DeiT backbones. Only generic/optional parameters retain defaults `(patch_size=4, mlp_ratio=4.0, qkv_bias=True, dropout rates, patch_norm=True, data_format=None, dtype=None)`. Docstrings now include "Defaults to X." for all keyword arguments with defaults.
Thank you for this feedback!

AdonaiVera committed:
Address feedback from sachinprasadhs, remove type annotations, clear presets, use isinstance(int)

AdonaiVera commented on keras_hub/src/models/swin_transformer/swin_transformer_backbone_test.py:
```diff
@@ -0,0 +1,58 @@
+            input_data=self.input_data,
+            expected_output_shape=(1, 49, 768),
+            run_mixed_precision_check=False,
+            run_quantization_check=False,
```
Same, solved it!

AdonaiVera commented on tools/checkpoint_conversion/convert_swin_transformer_checkpoints.py:
```diff
@@ -0,0 +1,255 @@
+    img_np = np.array(image_resized).astype("float32") / 255.0
+    img_np = (img_np - np.array([0.485, 0.456, 0.406])) / np.array(
+        [0.229, 0.224, 0.225]
+    )
```
This is a good one, thank you!

AdonaiVera committed:
Applied rabbit feedback and style improvement

AdonaiVera committed:
Fix attention mask slicing for shifted windows

AdonaiVera committed:
Fixed PatchEmbedding LayerNorm epsilon

AdonaiVera committed:
Add classifier, preprocessor, and converter task models

AdonaiVera commented:
Thank you so much for taking the time to review the full implementation in detail @sachinprasadhs . I applied your feedback as well as Gemini’s feedback.

Regarding your question:
After debugging layer by layer, I found that the issue was actually very simple but had a significant impact. The PatchEmbedding LayerNorm was using the Keras default epsilon (1e-3) instead of 1e-5. This caused output drift compared to the reference HF model. After fixing this, the outputs now match within a mean absolute difference of 2e-6. I also updated the [Colab notebook](https://colab.research.google.com/drive/1hw-HKrEZT4nqoN-poY6G2NxE1xrw6t04?usp=sharing), where you can see the results.
```
======================================================================
OUTPUT COMPARISON
======================================================================
Shape:                    HF=(1, 49, 768), Keras=(1, 49, 768)
Mean absolute difference: 0.000002
Max absolute difference:  0.000025
Standard deviation:       0.000002
Relative error:           0.00%

Sample values (first token, first 10 dims):
  HuggingFace: [ 0.33715773 -0.03832768  0.10057538  0.48289573 -0.5981863   0.03941139
  0.5929307   0.00331916 -1.693208    0.23169239]
  Keras:       [ 0.33715692 -0.0383272   0.10057551  0.48289585 -0.59818614  0.03941049
  0.5929307   0.00331995 -1.693208    0.231692  ]
```

And about your second and third question, HF Swin model includes classification head weights (`SwinForImageClassification`), and I just added the full task model stack here too so users can easily fine-tune or build classifiers on top of the backbone. I added:

* **`SwinTransformerImageClassifier`**: classification head (GAP over sequence + Dense).
* **`SwinTransformerImageClassifierPreprocessor`**: wraps the image converter for the classification pipeline.
* **`SwinTransformerImageConverter`**: handles resizing and normalization.

I also updated the backbone to store layer references (`self.patch_embedding`, `self.stages`, `self.norm`) so weights can be accessed programmatically, which is required by the conversion script. The final `LayerNorm` was renamed from `output_norm` to `norm` .

The validation notebook was updated to test the full pipeline (converter, preprocessor, classifier, `predict()`, serialization). Everything passes, and the backbone still matches HF with a **mean diff of 2e-6**.

AdonaiVera commented on keras_hub/src/models/swin_transformer/swin_transformer_layers.py:
```diff
@@ -0,0 +1,807 @@
+                    )
+                    img_mask = ops.slice_update(
+                        img_mask, [0, h_start, w_start, 0], mask_slice
+                    )
```
Thanks for flagging this! However, `ops.slice_update` is a valid Keras op, it's part of `keras.ops.core` and is documented [here](https://keras.io/api/ops/core/#sliceupdate-function). It works across all backends (TensorFlow, JAX, PyTorch) and is widely used throughout the keras-hub codebase (50+ files including attention layers, samplers, etc.). Our validation notebook confirms it runs correctly. But, if the mainteners think that is valid to apply this, happy to do it, for now I'm going to mark as solved it!

AdonaiVera commented on keras_hub/src/models/swin_transformer/swin_transformer_layers.py:
```diff
@@ -0,0 +1,807 @@
+        in_features,
+        hidden_features=None,
+        out_features=None,
+        act_layer=keras.activations.relu,
```
Good catch, thank you!

AdonaiVera commented on keras_hub/src/models/swin_transformer/swin_transformer_layers.py:
```diff
@@ -0,0 +1,807 @@
+                    dtype="float32",
+                )
+                * -100.0
+            )
```
Good thought, but -100.0 is intentionally aligned with the HF Swin implementation which uses the same value. It's actually safer for float16 than -1e9, float16 max is ~65504, so -1e9 would overflow to -inf and could cause NaN issues. exp(-100) is already effectively zero in both float32 and float16.

AdonaiVera committed:
Applied rabbit suggestion and move sublayers to init

AdonaiVera commented:
@gemini-code-assist review

AdonaiVera committed:
Remove redundant patch_norm from PatchEmbedding layer

AdonaiVera committed:
Fix JAX tracing and channels_first test

AdonaiVera committed:
the file say that is autogen but its keep failing in the build

AdonaiVera committed:
Merge branch 'master' into add-swin-transformer-model

AdonaiVera commented:
Hi @sachinprasadhs, sorry for pinging again! I would love to see this model in the ecosystem because it's the backbone to other cool models we can integrate with too

sachinprasadhs commented:
@AdonaiVera Apologies for the delay, this is on me, will take a look into it today. Thanks

sachinprasadhs commented on keras_hub/src/models/swin_transformer/__init__.py:
```diff
@@ -0,0 +1,18 @@
+)
+from keras_hub.src.models.swin_transformer.swin_transformer_image_converter import (
+    SwinTransformerImageConverter,
+)
```
No need of these, Just backlbone and registering preset would be enough.

sachinprasadhs commented on keras_hub/src/models/swin_transformer/swin_transformer_backbone_test.py:
```diff
@@ -0,0 +1,37 @@
+            init_kwargs=self.init_kwargs,
+            input_data=self.input_data,
+            expected_output_shape=(1, 16, 64),
+            run_data_format_check=False,
```
Will it be possible to enable data_format_check or is it incompatible for this model due to it's requirements?

AdonaiVera replied:
Good question. Right now it’s basically incompatible. Swin always outputs `(B, L, C)`, it works on flattened tokens, so we don’t get a spatial tensor like `(B, H, W, C)` at the end. The data format check expects a 4D tensor, so it will always fail with the current output.

We could make it work by adding an optional flag to reshape the output back to spatial (like HuggingFace does), but that’s not implemented yet. I’d suggest adding that in a follow-up PR if needed.

sachinprasadhs replied:
Makes sense, thanks for clarifying.

sachinprasadhs commented on tools/checkpoint_conversion/convert_swin_transformer_checkpoints.py:
```diff
@@ -0,0 +1,255 @@
+    final_w = get_hf_weight("layernorm.weight")
+    final_b = get_hf_weight("layernorm.bias")
+    if final_w is not None:
+        backbone.norm.set_weights([final_w, final_b])
```
This code can be moved under keras_hub/src/utils/transformers and in this converter file, you can do on the fly conversion from HF to convert checkpoints and validate.

AdonaiVera replied:
Sure, I'll do this

AdonaiVera replied:
I added the new converter in keras_hub/src/utils/transformers/convert_swin_transformer.py, following the same pattern as the other converters like convert_vit.py.

The old script in tools/checkpoint_conversion/convert_swin_transformer_checkpoints.py is still there. Should I remove it and keep everything in the new converter file? @sachinprasadhs

sachinprasadhs replied:
Yes, remove it. You should load it with from_preset("hf"//) This will do on the fly weights conversion using convert_swin_transformer.py

sachinprasadhs replied:
Make sure to load and save preprocessor also from on the fly conversion after you add `load_image_converter_config` to swin_transformer_presets.py
Right now, you are only saving model configs.

AdonaiVera replied:
Yes! I added `load_image_converter_config`. Now It reads `preprocessor_config.json` from the HF preset and maps the `scale`, `offset`, `image_size`, and `interpolation` values to the KerasHub format, so `from_preset("hf://...")` now sets up the image converter automatically too.

And I also cleaned up the checkpoint script by removing the manual `convert_model` and `convert_backbone_weights` flow and switching everything to `from_preset("hf://...")`. The script now saves the preprocessor together with the backbone as well.

sachinprasadhs commented on keras_hub/src/models/swin_transformer/swin_transformer_image_classifier.py:
```diff
@@ -0,0 +1,133 @@
+        x = self.output_dropout(x)
+        outputs = self.output_dense(x)
+
+        Task.__init__(
+            self,
+            inputs=inputs,
+            outputs=outputs,
+            **kwargs,
+        )
```
I belive this is the same reason as ViTImageClassifier?

AdonaiVera replied:
Yes exactly @sachinprasadhs. 
Do you want me to add a comment like in ViTImageClassifier for consistency? Happy to add it 👍

sachinprasadhs replied:
Yes please

AdonaiVera replied:
Perfect! I just did 💪

sachinprasadhs reviewed (commented):
Thanks, Looks good overall. Attached colab results also looks fantastic!
Added few small comments and clarification.
Also, in colab, you you add the prediction of the input. Right now I believe only logits and params comparision is happening and top 5 prediction with random weights.

AdonaiVera committed:
Add Swin converter and register HuggingFace preset loader

AdonaiVera commented:
Thank you for the feedback! @sachinprasadhs , 
Okey, I added `convert_swin_transformer.py` under `keras_hub/src/utils/transformers/`, following the same pattern as `convert_vit.py`. And I validated it, we match HuggingFace with ~1e-6 mean absolute difference.

The only question that I have now, its that the old script in `tools/checkpoint_conversion/` is still there. Should I remove it and keep everything in the new converter?

AdonaiVera commented:
Quick note, because I removed those classes from `__init__.py`, now `api-gen` fails since it regenerates the public API from exports. And its failling one of the tests, is that okey @sachinprasadhs ?

AdonaiVera commented:
And as you suggested, I added a new section in the Colab that loads SwinForImageClassification with pretrained ImageNet weights, converts the head, and compares top-5 predictions with HuggingFace.

Results match perfectly, mean difference is ~0.0, same top class (“tabby cat” at 37.47%). [Here](https://colab.research.google.com/drive/1hw-HKrEZT4nqoN-poY6G2NxE1xrw6t04?usp=sharing) is the Colab. 

```
======================================================================
REAL PREDICTIONS WITH PRETRAINED IMAGENET WEIGHTS
======================================================================
Loading SwinForImageClassification...
Loading weights: 100%
 233/233 [00:00<00:00, 688.89it/s, Materializing param=swin.layernorm.weight]
✅ Loaded — 1000 ImageNet classes
Building Keras classifier with pretrained head...
✅ Head weights converted

Top 5 Predictions (pretrained ImageNet weights):
Rank  Class                           HuggingFace        Keras
--------------------------------------------------------------
1     tabby, tabby cat                     37.47%       37.47%
2     tiger cat                            14.19%       14.19%
3     Egyptian cat                         11.66%       11.66%
4     remote control, remote                5.93%        5.93%
5     crutch                                1.80%        1.80%

Mean probability difference: 0.000000
✅ Predictions match HuggingFace!
```

We are ready for review @sachinprasadhs , let me know what do you think :rocket:

sachinprasadhs commented on keras_hub/src/utils/transformers/convert_swin_transformer.py:
```diff
@@ -0,0 +1,102 @@
+)
+
+backbone_cls = SwinTransformerBackbone
+
```
This converter should also have `load_image_converter_config` to load and convert preprocessor_config from the HF config files.
I know we don't have for many of the models, which is a implementation gap for on the fly conversion model, this would leave image_converter as None.
You need to fetch those values and convert it to Keras Hub convention like scale, offset etc.

So that when you call from_preset, the model would already apply preprocessor config from these.

AdonaiVera replied:
Done, more context in the other reply @sachinprasadhs !

AdonaiVera committed:
Added comment about functional model

sachinprasadhs commented:
> Quick note, because I removed those classes from `__init__.py`, now `api-gen` fails since it regenerates the public API from exports. And its failling one of the tests, is that okey @sachinprasadhs ?

API registering will be handled though `@keras_hub_export` so the model __init__ file is just to perform  global preset registry for vit presets when the ViT module is imported.

sachinprasadhs reviewed (commented):
Thanks, the notebook result looks good. 
Although, the orignal image size mentioned in preprocessor_config is 384 vs 224 is what you are resizing in the colab.

I have added few comments about preprocessor configs.

The reason we have tools/checkpoint_conversion is to validate the numerics, output and then generate the checkpoints which will be uploaded to kaggle using this script.
This script uses convert_swin_transformer to do on the fly conversion.

AdonaiVera committed:
Add load_image_converter_config and refactor Swin checkpoint

AdonaiVera commented:
Hi @sachinprasadhs, sorry to answer late, very busy week! Trying to keep working on this during the weekend :muscle: 

I applied your suggestion and I updated the notebook with a few improvements:

* Replaced the manual weight conversion code with a single `from_preset("hf://...")` call. This also validates the full on-the-fly conversion flow end to end.
* The image converter is now loaded directly from the preset as well, validating that `load_image_converter_config` correctly reads the `scale`, `offset`, and `image_size` values from the HF preprocessor config.
* Keep geeting good results with the real predictions section using `SwinForImageClassification` with pretrained ImageNet weights. The top-5 predictions keep matching the HF exactly.


Ready for review :muscle: :ship:

sachinprasadhs reviewed (commented):
Thanks, Looks good. Just one change. 
In the conversion script add assert statement for parameters count matching and a print statement.
And fix the code format issue. Rest all looks good. Thank you!

AdonaiVera committed:
Fix ruff format and add param count assert

sachinprasadhs commented:
Still it's failing.. could you please run `pre-commit run --all-files` and make sure it passes before you commmit changes.

AdonaiVera committed:
Pre-comit al the files

AdonaiVera commented:
Perfect @sachinprasadhs, thank you! I already runned it, and I think its ready for review :muscle:

sachinprasadhs reviewed (approved):
LGTM, Thank you!

AdonaiVera commented:
Amazing, thanks @sachinprasadhs ! Excited to see this model in the hub 🎉

sachinprasadhs commented:
Great, Thank you for your contribution!
